src/test/scala/is/hail/distributedmatrix/HailBlockMatrixSuite.scala

```diff
@@ -0,0 +1,522 @@
+    val numRows = 100
+    val numCols = 100
+    val local = BDM.rand[Double](numRows, numCols)
+    assert(local === HailBlockMatrix.from(sc, local, numRows - 1).toLocalMatrix())
+  }
+
+  @Test
+  def readWriteIdentityTrivial() {
+    val actual = toBM(Seq(
+      Array[Double](1,2,3,4),
+      Array[Double](5,6,7,8),
+      Array[Double](9,10,11,12),
+      Array[Double](13,14,15,16)))
+
+    val fname = tmpDir.createTempFile("test")
+    HailBlockMatrix.write(actual, fname)
+    assert(actual.toLocalMatrix() == HailBlockMatrix.read(hc, fname).toLocalMatrix())
+  }
+
+  @Test
+  def readWriteIdentityRandom() {
+    forAll(blockMatrixGen) { (bm: HailBlockMatrix) =>
+      val fname = tmpDir.createTempFile("test")
+      HailBlockMatrix.write(bm, fname)
+      assert(bm.toLocalMatrix() == HailBlockMatrix.read(hc, fname).toLocalMatrix())
+      true
+    }.check()
+  }
+
+  @Test
+  def transpose() {
+    forAll(blockMatrixGen) { (bm: HailBlockMatrix) =>
+      val transposed = bm.toLocalMatrix().t
+      assert(transposed.rows == bm.cols)
+      assert(transposed.cols == bm.rows)
+      assert(transposed === bm.t.toLocalMatrix())
+      true
+    }.check()
+  }
+
+  @Test
+  def doubleTransposeIsIdentity() {
+    forAll(blockMatrixGen) { (bm: HailBlockMatrix) =>
+      val bmt = bm.t.cache()
+      val bmtt = bm.t.t.cache()
+      assert(bmtt.rows == bm.rows)
+      assert(bmtt.cols == bm.cols)
+      assert(arrayEqualNaNEqualsNaN(bmtt.toLocalMatrix().toArray, bm.toLocalMatrix().toArray))
+      assert(arrayEqualNaNEqualsNaN((bmt * bmtt).toLocalMatrix().toArray, (bmt * bm).toLocalMatrix().toArray))
+      true
+    }.check()
+  }
+
+  @Test
+  def cachedOpsOK() {
+    forAll(twoMultipliableBlockMatrices) { case (a: HailBlockMatrix, b: HailBlockMatrix) =>
+      a.cache()
+      b.cache()
+
+      val truth = (a * b).toLocalMatrix()
+      val expected = a.toLocalMatrix() * b.toLocalMatrix()
+
+      if (!arrayEqualNaNEqualsNaN(truth.toArray, expected.toArray)) {
+        println(s"${a.toLocalMatrix()}")
+        println(s"${b.toLocalMatrix()}")
+        println(s"$truth != $expected")
+        assert(false)
+      }
+
+      if (!arrayEqualNaNEqualsNaN(a.t.cache().t.toLocalMatrix().toArray, a.toLocalMatrix.toArray)) {
+        println(s"${a.t.cache().t.toLocalMatrix()}")
+        println(s"${a.toLocalMatrix()}")
+        println(s"$truth != $expected")
+        assert(false)
+      }
+
+      true
+    }.check()
+  }
+
+  @Test
+  def toIRMToHBMIdentity() {
+    forAll(blockMatrixGen) { (bm : HailBlockMatrix) =>
+      val roundtrip = bm.toIndexedRowMatrix().toHailBlockMatrix(bm.blockSize)
+
+      val roundtriplm = roundtrip.toLocalMatrix()
+      val bmlm = bm.toLocalMatrix()
+
+      if (roundtriplm != bmlm) {
+        println(roundtriplm)
+        println(bmlm)
+        assert(false)
+      }
+
+      true
+    }.check()
+  }
+
+  @Test
+  def map2RespectsTransposition() {
+    val m = new BDM[Double](4,2, Array[Double](
```

Comment:
this one is 2,4. the next one is 4,2

Reply:
consider making `m` and `mt` external lazy vals, since you use the same ones in five tests.

Reply:
Spaces after commas, here and below